[custom-devices] general improvements, add custom-devices subcommand, better error handling

ardera opened the pull request:
- generally improves the custom devices feature
- adds flutter subcommand for custom-devices management:
  - `flutter custom-devices list` lists the configured custom devices
    - can be used to check for errors
    - will also output the path of the config
  - `flutter custom-devices reset`
    - resets the configured custom devices config to default & creates a backup
  - `flutter custom-devices add`
    - either supply a json-formatted string as an argument to the `--json` parameter
    - or configure using an interactive, guided device setup
    - that guided setup will by default setup a `ssh` device, i.e. it won't ask for the values of every custom device config option, but rather for things like the ssh username & hostname and generally provide some good defaults for ssh devices
  - `flutter custom-devices delete`
    - deletes a custom-devices file based on the device id
- better error checking when loading the config & better error messages
- if anything goes wrong while loading a `CustomDeviceConfig`, a `JsonRevivalException` will be thrown with an error message describing the error (like `Expected 'pingSuccessRegex' to be null or string-ified regex.`)
- that error will also be catched and expanded a bit by `CustomDevicesConfig.devices` so the outputted error message may look like this: `Could not load custom device from config index 0: Expected 'pingSuccessRegex' to be null or string-ified regex.`

fixes #79406

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ardera committed:
better config error handling
- make CustomDeviceConfig throw `JsonRevivalException` when it couldn't load the config.
- make CustomDevicesConfig log those JsonRevivalExceptions
- make CustomDevicesConfig log other config loading errors
- add tests for CustomDevicesConfig for testing those errors
- implement more dataclass-like methods for CustomDeviceConfig
- make CustomDeviceConfig immutable
- improve default pingSuccessRegex
- move some common stuff from the custom device tests into a custom_devices_common.dart
- fix some formatting issues

ardera committed:
- add (stub) custom devices command
- hookup track-widget-creation and tree-shake-icons options
- hookup runtime engine options

ardera committed:
use relative import

ardera committed:
dont log error if ping command fails (since that happens all the time when the device is unreachable)

ardera committed:
document the custom-devices reset command

ardera committed:
implement some of the custom-devices subcommands

ardera committed:
improvements to custom-devices add command

ardera committed:
ssh flag for add command, some error handling improvements

ardera committed:
- some config refactors
- more error handling
- remove deleteFileOnFormatException config ctor argument and instead introduce Config.managed ctor
- only execute custom devices subcommands when feature is enabled, provide hints to enable it

ardera committed:
Move config path hint in custom-devices command description into seperate paragraph

ardera committed:
fix typo

ardera committed:
make the usages nicer to read

ardera committed:
fix logic in checkFeatureEnabled

ardera committed:
fix logic in usage of checkFeatureEnabled

ardera committed:
- add a lot of tests
- allow overriding the print function of the custom-devices command so we can test the usage
- similiar to how it works with `flutter build`, make an invocation of `flutter custom-devices` print the usage
- add some docs to terminal.dart for clarification
- rename JsonRevivalException to CustomDevicesRevivalException
- implement checking the config (execute the commands and see if they work)
- make some methods public that are needed for executing the commands, maybe create a new class CustomDeviceCommandRunner in the future that has the sole purpose of running the command

ardera committed:
- add some new tests
- fix some tests
- for `custom-devices delete` command: use -d option for specifying the device to be deleted instead of --id
- add platform argument to CustomDevicesConfig so ensureInitialized will put an example for the right platform

ardera committed:
document some CustomDeviceAppSession stuff (and whether args accept null or not)

use tryGetDevices() instead of possibly exception-throwing devices getter inside CustomDevice

ardera committed:
fix capitalization in cmdline argument help

ardera committed:
use double-quotes instead of backticks in commandline output

ardera committed:
use double-quotes instead of backticks in commandline output for custom-devices --json option

ardera committed:
implement preferredStyle getter for FakeTerminal,
use double quotes in some instances

ardera committed:
fix some style issues

ardera committed:
add some more docs

HidenoriMatsubayashi commented:
Great work! I have one feedback about `custom-devices.schema.json`. There may be some reason, but I thought `enabled` is more intuitive than `disabled` ｗhen I first saw it.

```
          "disabled": {
            "description": ~,
            "type": "boolean"
            "type": "boolean",
            "default": false
          },
```

ardera commented:
> Great work! I have one feedback about `custom-devices.schema.json`. There may be some reason, but I thought `enabled` is more intuitive than `disabled` ｗhen I first saw it.

Thanks! Yeah, the style guide also [mentions this](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-double-negatives-in-apis), I chose to ignore it because I thought it'd be weird for a field to default to "true" if omitted, but you're probably right I should've called it `enabled` instead. Also leads to some awkward double negatives in the code

I'll change it 👍

ardera committed:
- rename "disabled" to "enabled" (and invert its value and make not required)
- add target platform to the schema (currently only linux-arm64 and linux-x64 are allowed)
- fix some mismatching " and `
- fix duplicate test
- fix test naming in one instance

ardera committed:
close stream terminal input stream later

ardera committed:
close stream terminal input stream later
add test for custom device returning correct target platform

ardera committed:
fix unawaited future

ardera commented:
this should be ready to be reviewed now @jonahwilliams (just fyi)

jonahwilliams commented:
Awesome, will take a look at this tomorrow!

jonahwilliams commented on packages/flutter_tools/test/general.shard/custom_devices/custom_devices_config_test.dart:
```diff
@@ -0,0 +1,267 @@
+    expect(() => customDevicesConfig.devices, throwsA(const CustomDeviceRevivalException(msg)));
+    expect(logger.errorText, contains(msg));
+  });
+}
```
```suggestion
}

```

jonahwilliams reviewed (commented):
Still looking through this one, just some initial feedback

ardera commented on packages/flutter_tools/lib/src/commands/custom_devices.dart:
```diff
@@ -0,0 +1,829 @@
+    List<CustomDeviceConfig> devices;
+    try {
+      devices = customDevicesConfig.devices;
+    } on Exception catch (_) {
```
I think I did those catch clauses because otherwise the analyzer was complaining for some reason, I'll try again

ardera commented on packages/flutter_tools/lib/src/commands/custom_devices.dart:
```diff
@@ -0,0 +1,829 @@
+        "By default, this won't check whether the passed in config actually "
+        'works (only if it is valid). To make sure the config works use the '
+        '"--check" option.',
+      valueHelp: 'JSON config',
```
A config with a null `runDebug` command would be invalid, a config with `["dawdawaw"]` as the `runDebug` command would be valid, but wouldn't work

So you can use the `--check` option to make sure it the config actually works, which will execute some of the commands of the config with dummy arguments and see if they succeed.

But if you have to ask to understand it, that's probably a good sign the help isn't helpful enough 😄

ardera replied:
I updated the help, is it better?

jonahwilliams replied:
ahh I see. In this case, it is better to either leave off the value help or provide an actual example of the data you would pass : `valueHelp: '{...};`. That might be too much for the CLI help option

ardera replied:
Okay, I changed the `valueHelp` to `{"id": "pi", ...}` now

ardera commented on packages/flutter_tools/lib/src/commands/custom_devices.dart:
```diff
@@ -0,0 +1,829 @@
+
+      try {
+        // find a random port we can forward
+        final int port = Random().nextInt(65536 - 49152) + 49152;
```
oh nice! I tried to find one but couldn't

ardera committed:
Update packages/flutter_tools/lib/src/commands/custom_devices.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/commands/custom_devices.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/commands/custom_devices.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
don't use 'else' if 'if' returns/throws unconditionally
- use ErrorHandlingFilesystem to delete file
- use OperatingSystemUtils to find a free port
- improve the custom-devices add help a bit

ardera committed:
Update packages/flutter_tools/test/general.shard/custom_devices/custom_devices_config_test.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera commented on packages/flutter_tools/lib/src/commands/custom_devices.dart:
```diff
@@ -0,0 +1,829 @@
+
+    await backup();
+
+    await fileSystem.file(customDevicesConfig.configPath).delete();
```
Yep but I'm not sure reporting the success is necessary, all that `reset` should do is ensure the user has a blank config file after the command is done, regardless of the state before

ardera committed:
- make CustomDevicesCommandBase.backup work even when the config file doesn't exist, and return false when that's the case
- in test, match config file contents after executing custom-devices reset
- add test for correct custom-devices reset message

jonahwilliams commented on packages/flutter_tools/lib/src/commands/custom_devices.dart:
```diff
@@ -0,0 +1,849 @@
+  /// if necessary.
+  @protected
+  Future<bool> backup() async {
+    final File configFile = fileSystem.file(customDevicesConfig.configPath);
+    if (await configFile.exists()) {
+      await configFile.copy(configBackupPath);
+      return true;
+    }
+    return false;
```
nit: prefer sync io

ardera replied:
done

jonahwilliams commented on packages/flutter_tools/lib/src/commands/custom_devices.dart:
```diff
@@ -0,0 +1,849 @@
+  Future<FlutterCommandResult> runCommand() async {
+    checkFeatureEnabled();
+
+    if (stringArg(_kJson) != null) {
+      return runNonInteractively();
+    } else if (boolArg(_kSsh) == true) {
+      return runInteractivelySsh();
+    } else {
+      throw FallThroughError();
+    }
```
```suggestion
    if (stringArg(_kJson) != null) {
      return runNonInteractively();
    }
    if (boolArg(_kSsh) == true) {
      return runInteractivelySsh();
    } 
    throw FallThroughError();
```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart:
```diff
@@ -61,26 +223,91 @@ class CustomDeviceConfig {
+  /// development machine. Examples for different platforms may be different
+  /// because for example the ping command is different on Windows or Linux/macOS.
+  static CustomDeviceConfig getExampleForPlatform(Platform platform) {
+    if (platform.isWindows) {
+      return exampleWindows;
+    } else if (platform.isLinux || platform.isMacOS) {
+      return exampleUnix;
+    } else {
+      throw FallThroughError();
+    }
```
```suggestion
    if (platform.isWindows) {
      return exampleWindows;
    }
    if (platform.isLinux || platform.isMacOS) {
      return exampleUnix;
    }
    throw FallThroughError();
```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart:
```diff
@@ -91,28 +318,159 @@ class CustomDeviceConfig {
+    );
   }
 
   dynamic toJson() {
```
```suggestion
  Object toJson() {
```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart:
```diff
@@ -161,4 +522,60 @@ class CustomDeviceConfig {
+  }
+
+  @override
+  int get hashCode {
```
I think we might already have a quiver dep for hashcode you should consider instead

ardera replied:
you're right, using that instead

ardera replied:
btw I couldn't use this ([`hashObjects`](https://pub.dev/documentation/quiver/latest/quiver.core/hashObjects.html)) because the quiver flutter uses is still not null-migrated, so I can't use it inside `custom_device_config.dart`

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart:
```diff
@@ -84,28 +102,123 @@ class CustomDevicesConfig {
+  List<CustomDeviceConfig> tryGetDevices() {
+    try {
+      return devices;
+    } on Exception catch (_) {
```
```suggestion
    } on Exception {
```

jonahwilliams commented on packages/flutter_tools/lib/src/globals_null_migrated.dart:
```diff
@@ -211,3 +212,5 @@ FlutterProjectFactory get projectFactory {
   );
 }
+
+CustomDevicesConfig get customDevicesConfig => context.get<CustomDevicesConfig>()!;
```
```suggestion
CustomDevicesConfig get customDevicesConfig => context.get<CustomDevicesConfig>()!;

```

jonahwilliams reviewed (commented):
FallThroughError should be an `Exception` type if it is intended to be caught.

ardera committed:
Update packages/flutter_tools/lib/src/commands/custom_devices.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/lib/src/globals_null_migrated.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
- prefer sync io
- make config backup function synchronous

ardera commented:
> FallThroughError should be an `Exception` type if it is intended to be caught.

I see, in the one instance where I throw it, it's not intended to be caught (I mean I don't know what to do when the command is run on something other than macOS, windows or linux)

Also I did something wrong while rebasing (somehow I rebased master on top of custom-device-improvements instead of the other way around), all the master commits showed up in this PR and somehow the bot requested two authors of the commits for review. Maybe you can remove the two from the reviewer list? @jonahwilliams

ardera committed:
fix trailing whitespace

zanderso commented:
@ardera Is this ready for another review pass?

FYI @eseidel

ardera commented:
@zanderso yep it's ready

jonahwilliams reviewed (approved):
LGTM

zanderso commented:
@ardera I would love to see an article in our Wiki (to start) with a tutorial on how to use this.

ardera commented:
> @ardera I would love to see an article in our Wiki (to start) with a tutorial on how to use this.

sure, good idea!
